src/rt/region/region_rc.h

```diff
@@ -0,0 +1,408 @@
+
+    static const Descriptor* desc()
+    {
+      static constexpr Descriptor desc = {
+        vsizeof<RegionRc>, nullptr, nullptr, nullptr};
+
+      return &desc;
+    }
+
+  public:
+    inline static RegionRc* get(Object* o)
+    {
+      assert(o->debug_is_iso());
+      assert(is_rc_region(o->get_region()));
+      return (RegionRc*)o->get_region();
+    }
+
+    inline static bool is_rc_region(Object* o)
+    {
+      return o->is_type(desc());
+    }
+
+    /**
+     * Creates a new rc region by allocating Object `o` of type `desc`. The
+     * object is initialised as the Iso object for that region, and points to a
+     * newly created Region metadata object. Returns a pointer to `o`.
+     *
+     * The default template parameter `size = 0` is to avoid writing two
+     * definitions which differ only in one line. This overload works because
+     * every object must contain a descriptor, so 0 is not a valid size.
+     **/
+    template<size_t size = 0>
+    static Object* create(Alloc& alloc, const Descriptor* desc)
+    {
+      void* p = alloc.alloc<vsizeof<RegionRc>>();
+      Object* o = Object::register_object(p, RegionRc::desc());
+      auto reg = new (o) RegionRc();
+      reg->use_memory(desc->size);
+
+      if constexpr (size == 0)
+        p = alloc.alloc(desc->size);
+      else
+        p = alloc.alloc<size>();
+      o = Object::register_object(p, desc);
+
+      reg->init_next(o);
+      o->init_iso();
+      o->set_region(reg);
+
+      assert(Object::debug_is_aligned(o));
+      return o;
+    }
+
+    /**
+     * Allocates an object `o` of type `desc` in the region represented by the
+     * Iso object `in`, and adds it to the appropriate ring. Returns a pointer
+     * to `o`.
+     *
+     * The default template parameter `size = 0` is to avoid writing two
+     * definitions which differ only in one line. This overload works because
+     * every object must contain a descriptor, so 0 is not a valid size.
+     **/
+    template<size_t size = 0>
+    static Object* alloc(Alloc& alloc, Object* in, const Descriptor* desc)
+    {
+      assert((size == 0) || (size == desc->size));
+      RegionRc* reg = get(in);
+
+      assert(reg != nullptr);
+
+      void* p = nullptr;
+      if constexpr (size == 0)
+        p = alloc.alloc(desc->size);
+      else
+        p = alloc.alloc<size>();
+
+      auto o = (Object*)Object::register_object(p, desc);
+      assert(Object::debug_is_aligned(o));
+
+      reg->track_object(o, alloc);
+
+      // GC heuristics.
+      reg->use_memory(desc->size);
+      return o;
+    }
+
+    static void incref(Object* o)
+    {
+      RefCount* rc = o->get_ref_count();
+      rc->metadata += 1;
+    }
+
+    static bool decref(Alloc& alloc, Object* o, Object* in)
+    {
+      if (decref_inner(o))
+      {
+        dealloc_object(alloc, o, in);
+        return true;
+      }
+      return false;
+    }
+
+    static uintptr_t debug_get_ref_count(Object* o)
```

Comment:
Is this only to be used in "debug".  It looks safe to use in a non-debug case.  It should be concurrency safe, as this is purely single-threaded region code.

Reply:
I only used this for debuggin/test code so I prefixed it with debug. I suppose you're right that it might be a useful API for non-debug usage. Fixed here: 9120678b2abdd330de73e156be48caa5537e689a